async/overview.md

```diff
@@ -1,2 +1,216 @@
+```
+/// Assume we get a future string back from some API
+let futureString: Future<String> = ...
+futureString.always {
+    print("The future is complete!")
+}
+```
+> 笔记
+> 
+> 你可以根据需求为future添加尽可能多的回调
+
+### Wait
+你可以使用.wait（）同步等待future完成。由于future可能会失败，所以这个调用会抛出异常。
+ ```
+ /// Assume we get a future string back from some API
+let futureString: Future<String> = ...
+/// 一直阻塞到string操作完成
+let string = try futureString.wait()
+print(string) /// String
+ ```
+> 警告
+>
+> 不要在一个vapor 的路由或者控制器中使用这个方法，阅读`Blocking`来获取更多信息。
+
+## Promise
+大多数情况下，你是通过调用vapor的API's来获取一个需要操作的future。但是，在某些时候，你可能需要创建自己的promise。
+
+要创建promise，你需要访问`EventLoop`。Vapor中的所有容器都有一个可以使用的`eventLoop`属性。最常见的是，就是当前的`Request`。
+
+```
+/// Create a new promise for some string
+let promiseString = req.eventLoop.newPromise(String.self)
+print(promiseString) // Promise<String>
+print(promiseString.futureResult) // Future<String>
+/// Completes the associated future
+promiseString.succeed(result: "Hello")
+/// Fails the associated future
+promiseString.fail(error: ...)
+```
+
+> 信息
+>
+> 一个Promise对象只能够被完成一次，所有的后续完成操作都会被忽略。
+
+### 线程安全
+
+Promises可以从任何线程完成` (succeed(result:) / fail(error:)) `。这就是为什么promise需要一个事件循环来初始化。Promise确保完成操作在它的事件循环执行。
+
+## Event Loop （事件循环）
+
+它通常会为其运行的CPU中的每个内核创建一个事件循环在你的应用程序启动时。每个事件循环只有一个线程。如果你熟悉Node.js的事件循环，就会发现Vapor中的事件循环与Node.js中的非常相似。唯一的区别是Vapor可以在一个进程中运行多个事件循环，因为Swift支持多线程。
+
+每次客户端连接到你的服务器时，它都将被分配给其中一个事件循环。从这一点开始，服务器和该客户端之间的所有通信都将发生在同一个事件循环（并关联到这个事件循环的线程）上。
+
+事件循环负责跟踪每个连接的客户端的状态。如果有来自客户端的请求等待读取，则事件循环触发读取通知，导致数据被读取。一旦读取完整个请求，任何等待该请求数据的futures都将完成。
+
+### Worker
+有权访问事件循环的事物称为`Worker`。Vapor的每个容器都是一个worker。Vapor中最常用的容器是:
+* `Application`
+* `Request`
+* `Response`
+
+你可以在这些容器上使用`.eventLoo`p属性来访问事件循环。
+```
+print(app.eventLoop) // EventLoop
+```
+Vapor中有许多方法需要传递当前worker。它通常会有像`on: Worker`这样的标签。如果你处于路由封或控制器的闭包中，请传递当前`Request`或`Response`。如果你在启动应用程序时需要一个worker，请使用`Application`.
+
+### Blocking（阻塞）
+绝对关键的规则如下
+> 危险
+>
+> 切勿直接在事件循环中进行阻塞调用。
+
+一个阻塞调用的例子,类似于libc.sleep（_ :).
+```
+router.get("hello") { req in
+    /// 使当前事件循环阻塞5s
+    sleep(5)
+
+    ///当线程重新工作室返回一个简单字符串
+    return "Hello, world!"
+}
+```
+`sleep（_ :)`是一个用于在提供的秒数内阻止当前线程的命令。如果你直接在事件循环中阻塞工作，则在阻塞工作期间，事件循环将无法响应分配给它的任何其他客户端。换句话说，如果你在事件循环中进行睡眠（5），则连接到该事件循环的所有其他客户端（可能为数百或数千个）将被延迟至少5秒
+
+确保只有在后台才运行任何阻塞工作。当这项工作以非阻塞方式完成时，使用Promise通知事件循环。
+
+```
+router.get("hello") { req in
+    /// 创建一个空的promise
+    let promise = req.eventLoop.newPromise(Void.self)
+
+    /// 在后台线程执行
+    DispatchQueue.global() {
+        /// 使后台线程睡眠
+        /// 这不会影响到当前的事件循环
+        sleep(5)
+
+        /// 当阻塞工作完成时
+        /// promise和他关联的future会被完成
+        promise.succeed()
+    }
+
+    /// 等待 future完成
+    /// 将结果转为一个简单的字符
+    return promise.futureResult.transform(to: "Hello, world!")
+   
+}
+```
+并非所有的阻塞调用会像`sleep(:)`如此明显，当你怀疑你调用的方法可能被阻塞时，研究下这个方法本身或者询问相关的人。如果该函数正在执行磁盘或者网络IO并且使用的是一个同步API(无回调或者futures)，那么就会发生阻塞的情况。
+
+> 信息
+> 
+> 如果阻塞工作是应用程序的核心部分，则应该考虑使用BlockingIOThreadPool来控制你创建的阻塞工作的线程数。这可以避免你的阻塞工作正在完成时从饿死CPU时间开始的事件循环。
```

Comment:
原文：This will help you avoid starving your event loops from CPU time while blocking work is being done.

建议翻译：
当阻塞工作正在被完成时，这将避免你的事件循环长期获取不到 CPU 时间片。